tracer/src/Datadog.Trace/Agent/TraceSamplers/RareSampler.cs

```diff
@@ -0,0 +1,71 @@
+// <copyright file="RareSampler.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+
+namespace Datadog.Trace.Agent.TraceSamplers
+{
+    internal class RareSampler : ITraceSampler
+    {
+        private const string RareKey = "_dd.rare";
+        private readonly HashSet<StatsAggregationKey> _keys = new();
+
+        public bool Sample(ArraySegment<Span> trace)
+        {
+            bool priorityGreaterThan0 = trace.Array[trace.Offset].Context.TraceContext.SamplingPriority is int p && p > 0;
+            return priorityGreaterThan0 switch
+            {
+                true => HandlePriorityTrace(trace),
+                false => HandleTrace(trace)
+            };
+        }
+
+        private bool HandlePriorityTrace(ArraySegment<Span> trace)
+        {
+            for (int i = 0; i < trace.Count; i++)
+            {
+                var span = trace.Array[i + trace.Offset];
+                if (span.IsTopLevel || span.GetMetric(Tags.Measured) == 1.0 || span.GetMetric(Tags.PartialSnapshot) > 0)
+                {
+                    // Do not return immediately, as we might have multiple spans in this chunk that are rare
+                    var key = StatsAggregator.BuildKey(span);
+                    _keys.Add(key);
+                }
+            }
+
+            return false;
```

Comment:
Just to confirm, `HandlePriorityTrace` _always_ returns false? Also, `_keys` seems to have unbounded growth unless I'm mistaken?

Reply:
> Just to confirm, HandlePriorityTrace always returns false?

Yes, according to the agent logic we should return false when the trace has a positive sampling priority.

> Also, _keys seems to have unbounded growth unless I'm mistaken?

Yes, I didn't fix this in the last PR but now would be a good time to implement this. I'll think on the implementation and get back to this point.